src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributes.cs

```diff
@@ -0,0 +1,234 @@
+{
+    /// <summary>
+    /// Represents a collection of <see cref="TagHelperAttribute"/>s.
+    /// </summary>
+    public class TagHelperAttributes : ReadOnlyTagHelperAttributes<TagHelperAttribute>, IList<TagHelperAttribute>
+    {
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="TagHelperAttributes"/> with an empty collection.
+        /// </summary>
+        public TagHelperAttributes()
+            : base()
+        {
+        }
+
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="TagHelperAttributes"/> with the specified
+        /// <paramref name="attributes"/>.
+        /// </summary>
+        /// <param name="attributes">The collection to wrap.</param>
+        public TagHelperAttributes([NotNull] IEnumerable<TagHelperAttribute> attributes)
+            : base(attributes)
+        {
+        }
+
+        /// <inheritdoc />
+        /// <remarks><paramref name="value"/>'s <see cref="TagHelperAttribute.Name"/> cannot be <c>null</c>.</remarks>
+        public new TagHelperAttribute this[int index]
+        {
+            get
+            {
+                return base[index];
+            }
+            [param: NotNull]
+            set
+            {
+                ValidateAttribute(value, nameof(value));
+
+                Attributes[index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets the first <see cref="TagHelperAttribute"/> whose <see cref="TagHelperAttribute.Name"/> matches
+        /// the specified <paramref name="name"/>. When setting, replaces the first matching
+        /// <see cref="TagHelperAttribute"/> with the specified <paramref name="value"/> and removes any additional
+        /// matching <see cref="TagHelperAttribute"/>s. If a matching <see cref="TagHelperAttribute"/> is not found,
+        /// adds the specified <paramref name="value"/> to the end of the collection.
+        /// </summary>
+        /// <param name="name">
+        /// The <see cref="TagHelperAttribute.Name"/> of the <see cref="TagHelperAttribute"/> to get or set.
+        /// </param>
+        /// <returns>The first <see cref="TagHelperAttribute"/> whose <see cref="TagHelperAttribute.Name"/> matches the
+        /// specified <paramref name="name"/>.
+        /// </returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively. When setting,
+        /// <see cref="TagHelperAttribute"/>s <see cref="TagHelperAttribute.Name"/> must be <c>null</c> or
+        /// case-insensitively match the specified <paramref name="name"/>.</remarks>
+        /// <example>
+        /// var attributes = new TagHelperAttributes();
+        ///
+        /// // Will be converted to a TagHelperAttribute
+        /// attributes["name"] = "value";
+        ///
+        /// // TagHelperAttribute.Name must match the specified name.
+        /// attributes["name"] = new TagHelperAttribute("name", "value");
+        /// </example>
```

Comment:
`<example/>` delimits the section but doesn't change its formatting. use `<code/>` to wrap content.